python/pyiceberg/io/pyarrow.py

```diff
@@ -1025,3 +1039,335 @@ def map_key_partner(self, partner_map: Optional[pa.Array]) -> Optional[pa.Array]
+        return MetricsMode(MetricModeTypes.FULL)
+    else:
+        raise ValueError(f"Unsupported metrics mode: {mode}")
+
+
+@dataclass(frozen=True)
+class StatisticsCollector:
+    field_id: int
+    iceberg_type: PrimitiveType
+    mode: MetricsMode
+    column_name: str
+
+
+class PyArrowStatisticsCollector(PreOrderSchemaVisitor[List[StatisticsCollector]]):
+    _field_id: int = 0
+    _schema: Schema
+    _properties: Dict[str, str]
+
+    def __init__(self, schema: Schema, properties: Dict[str, str]):
+        self._schema = schema
+        self._properties = properties
+
+    def schema(self, schema: Schema, struct_result: Callable[[], List[StatisticsCollector]]) -> List[StatisticsCollector]:
+        return struct_result()
+
+    def struct(
+        self, struct: StructType, field_results: List[Callable[[], List[StatisticsCollector]]]
+    ) -> List[StatisticsCollector]:
+        return list(chain(*[result() for result in field_results]))
+
+    def field(self, field: NestedField, field_result: Callable[[], List[StatisticsCollector]]) -> List[StatisticsCollector]:
+        self._field_id = field.field_id
+        result = field_result()
+        return result
+
+    def list(self, list_type: ListType, element_result: Callable[[], List[StatisticsCollector]]) -> List[StatisticsCollector]:
+        self._field_id = list_type.element_id
+        return element_result()
+
+    def map(
+        self,
+        map_type: MapType,
+        key_result: Callable[[], List[StatisticsCollector]],
+        value_result: Callable[[], List[StatisticsCollector]],
+    ) -> List[StatisticsCollector]:
+        self._field_id = map_type.key_id
+        k = key_result()
+        self._field_id = map_type.value_id
+        v = value_result()
+        return k + v
+
+    def primitive(self, primitive: PrimitiveType) -> List[StatisticsCollector]:
+        column_name = self._schema.find_column_name(self._field_id)
+        if column_name is None:
+            raise ValueError(f"Column for field {self._field_id} not found")
+
+        metrics_mode = MetricsMode(MetricModeTypes.TRUNCATE, DEFAULT_TRUNCATION_LENGTH)
+
+        default_mode = self._properties.get(DEFAULT_METRICS_MODE_KEY)
+        if default_mode:
+            metrics_mode = match_metrics_mode(default_mode)
+
+        col_mode = self._properties.get(f"{COLUMN_METRICS_MODE_KEY_PREFIX}.{column_name}")
+        if col_mode:
+            metrics_mode = match_metrics_mode(col_mode)
+
+        if (
+            not (isinstance(primitive, StringType) or isinstance(primitive, BinaryType))
+            and metrics_mode.type == MetricModeTypes.TRUNCATE
+        ):
+            metrics_mode = MetricsMode(MetricModeTypes.FULL)
+
+        return [StatisticsCollector(field_id=self._field_id, iceberg_type=primitive, mode=metrics_mode, column_name=column_name)]
+
+
+def fill_parquet_file_metadata(
+    df: DataFile,
+    parquet_metadata: pq.FileMetaData,
+    file_size: int,
+    table_metadata: TableMetadata,
```

Comment:
I don't think that this should be passed in. All that's needed is the Parquet write schema and a map from field ID to metrics mode. There are good reasons not to require that bundled into `TableMetadata`:
1. It is safe and possible to write new data files with any older schema, so the current table schema may not be the write schema (this would currently break in that case)
2. Metrics modes can be pre-calculated to produce a map from field ID to mode. It makes more sense to do this in a common util method than to mix the code into the already complex `StatisticsCollector` visitor
3. Avoiding passing `TableMetadata` keeps this logic cleaner and forces some separation of concerns

Reply:
Yes, the first iteration had this separation of concerns: https://github.com/apache/iceberg/blob/ce8a5df5d07b51dbee03c7d8d88dcff99ffbde02/python/pyiceberg/utils/file_stats.py#L121 . We can split it up again.

Reply:
I agree that we shouldn't pass in the `TableMetadata` here. Since we know the write schema, we can just pass that one in.

> Metrics modes can be pre-calculated to produce a map from field ID to mode. It makes more sense to do this in a common util method than to mix the code into the already complex StatisticsCollector visitor

I'm not sure about this one, we know the order of the columns, so we can just rely on that one. We can also have a lookup, but I'm not sure that it makes it much simpler than keeping the mode next to the id and type.

Reply:
We could also just pass in the `stats_columns`:
```suggestion
    stats_columns: List[StatisticsCollector]
```